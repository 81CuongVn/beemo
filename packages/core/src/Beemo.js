/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { Pipeline, Tool } from 'boost';
import chalk from 'chalk';
import fs from 'fs-extra';
import path from 'path';
import ConfigureRoutine from './ConfigureRoutine';
import ExecuteRoutine from './ExecuteRoutine';
import SyncDotfilesRoutine from './SyncDotfilesRoutine';

import type { Event, Reporter } from 'boost';
import type { BeemoContext } from './types';
import type Driver from './Driver';

export default class Beemo {
  argv: string[];

  context: BeemoContext;

  tool: Tool<Driver, Reporter<Object>>;

  constructor(argv: string[]) {
    this.argv = argv;

    // eslint-disable-next-line global-require
    const { version } = require('../package.json');

    this.tool = new Tool({
      appName: 'beemo',
      configFolder: './configs',
      footer: `ðŸ¤–  Powered by Beemo v${version}`,
      pluginAlias: 'driver',
      scoped: true,
    }, argv);

    // Immediately load config and plugins
    this.tool.initialize();

    // Handle exit failures
    this.tool.on('exit', this.cleanUpOnFailure);
  }

  /**
   * Delete config files on a failure.
   */
  cleanUpOnFailure = (event: Event, code: number) => {
    if (code > 0 && this.context) {
      // Must not be async!
      this.context.configPaths.forEach((configPath) => {
        fs.removeSync(configPath);
      });
    }
  };

  /**
   * Validate the configuration module and return its absolute path.
   */
  getConfigModuleRoot(): string {
    const { config } = this.tool.config;

    this.tool.debug('Locating configuration module root');

    if (!config) {
      throw new Error(
        'Beemo requires a "beemo.config" property within your package.json. ' +
        'This property is the name of a module that houses your configuration files.',
      );
    }

    // Allow for local development
    if (config === '@local') {
      this.tool.debug(`Using ${chalk.yellow('@local')} configuration module`);

      return process.cwd();
    }

    const rootPath = path.join(process.cwd(), 'node_modules', config);

    if (!fs.existsSync(rootPath)) {
      throw new Error('Module defined in "beemo.config" could not be found.');
    }

    this.tool.debug(`Found configuration module root path: ${chalk.cyan(rootPath)}`);

    return rootPath;
  }

  /**
   * Execute all routines for the chosen driver.
   */
  executeDriver(driverName: string): Promise<*> {
    const { tool } = this;
    const configRoot = this.getConfigModuleRoot();
    const primaryDriver = tool.getPlugin(driverName);

    this.context = {
      // 0 node, 1 beemo, 2 driver
      args: this.argv.slice(3),
      argsObject: {},
      configPaths: [],
      configRoot,
      drivers: [primaryDriver],
      primaryDriver,
      root: this.tool.options.root,
    };

    this.tool.emit('run', [primaryDriver, this.context]);

    this.tool.debug(`Running with ${chalk.magenta(primaryDriver.name)} driver`);

    return new Pipeline(tool)
      .pipe(new ConfigureRoutine('configure', 'Generating configurations'))
      .pipe(new ExecuteRoutine('execute', 'Executing driver'))
      .run(driverName, this.context);
  }

  /**
   * Sync dotfiles from the configuration module.
   */
  syncDotfiles(): Promise<*> {
    this.tool.emit('sync');

    return new Pipeline(this.tool)
      .pipe(new SyncDotfilesRoutine('sync', 'Syncing dotfiles'))
      .run(this.getConfigModuleRoot());
  }
}
