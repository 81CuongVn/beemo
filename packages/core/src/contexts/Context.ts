/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Context as BaseContext } from '@boost/core';
import camelCase from 'lodash/camelCase';
import trim from 'lodash/trim';
import { Arguments, Argv } from '../types';

export interface ConfigPath {
  driver: string;
  path: string;
}

export default class Context extends BaseContext {
  args: Arguments;

  argv: Argv = [];

  configPaths: ConfigPath[] = [];

  moduleRoot: string = '';

  root: string = '';

  constructor(args: Arguments) {
    super();

    this.args = args;
  }

  /**
   * Add a positional argument to the argv list.
   */
  addArg(arg: string): this {
    this.args._.push(arg);
    this.argv.push(arg);

    return this;
  }

  /**
   * Add multiple positional arguments.
   */
  addArgs(args: string[]): this {
    args.forEach(arg => {
      this.addArg(arg);
    });

    return this;
  }

  /**
   * Add a config path for the defined driver.
   */
  addConfigPath(driverName: string, path: string): this {
    this.configPaths.push({
      driver: driverName,
      path,
    });

    return this;
  }

  /**
   * Add an option argument to both the args object and argv list.
   */
  addOption(arg: string, defaultValue: any = true, useEquals: boolean = false): this {
    const list = [];
    let option = arg;
    let value = defaultValue;

    if (option.includes('=')) {
      [option, value] = option.split('=');
      useEquals = true;
    }

    let name = trim(option, '-');

    if (name.startsWith('no-')) {
      name = name.slice(3);
      value = false;
    }

    this.args[name] = value;
    this.args[camelCase(name)] = value;

    if (typeof value === 'boolean' || !value) {
      list.push(option);
    } else if (useEquals) {
      list.push(`${option}=${value}`);
    } else {
      list.push(option, String(value));
    }

    this.argv.push(...list);

    return this;
  }

  /**
   * Add multiple boolean option arguments.
   */
  addOptions(args: string[]): this {
    args.forEach(arg => {
      this.addOption(arg);
    });

    return this;
  }

  /**
   * Find a configuration path by file name.
   */
  findConfigByName(name: string): ConfigPath | undefined {
    return this.configPaths.find(config => config.path.endsWith(name) || config.driver === name);
  }

  /**
   * Return an argument or option value by name, or a fallback value if not found.
   */
  getArg<T>(name: string, fallback: any = null): T {
    return this.args[name] || fallback;
  }
}
