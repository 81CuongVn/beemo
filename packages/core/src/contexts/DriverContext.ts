/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Context from './Context';
import Driver from '../Driver';
import { Arguments, Argv } from '../types';

export default class DriverContext extends Context {
  configPaths: string[] = [];

  driverName: string = '';

  drivers: Set<Driver<any>> = new Set();

  parallelArgv: Argv[] = [];

  primaryDriver: Driver<any>;

  workspaceRoot: string = '';

  workspaces: string[] = [];

  constructor(args: Arguments, driver: Driver<any>, parallelArgv: Argv[] = []) {
    super(args);

    this.driverName = driver.name;
    this.parallelArgv = parallelArgv;
    this.primaryDriver = driver;

    // Make the context available in the current driver
    driver.context = this;
  }

  /**
   * Add a driver as a dependency.
   */
  addDriverDependency(driver: Driver<any>): this {
    if (driver instanceof Driver) {
      this.drivers.add(driver);
    } else {
      throw new TypeError('Invalid driver. Must be an instance of `Driver`.');
    }

    return this;
  }

  /**
   * Add a parallel command with additional argv.
   */
  addParallelCommand(argv: Argv): this {
    this.parallelArgv.push(argv);

    return this;
  }

  /**
   * Find a configuration path by file name.
   */
  findConfigByName(fileName: string): string | undefined {
    return this.configPaths.find(path => path.endsWith(fileName));
  }
}
