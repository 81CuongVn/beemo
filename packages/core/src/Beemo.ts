/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import chalk from 'chalk';
import fs from 'fs-extra';
import path from 'path';
import { Argv as YargsApp } from 'yargs';
import { CLI, Pipeline, Tool } from '@boost/core';
import { bool, number, shape, string, Blueprint } from 'optimal';
import CleanupRoutine from './CleanupRoutine';
import ConfigureRoutine from './ConfigureRoutine';
import ExecuteDriverRoutine from './ExecuteDriverRoutine';
import ExecuteScriptRoutine from './ExecuteScriptRoutine';
import ScaffoldRoutine from './ScaffoldRoutine';
import Driver from './Driver';
import Context from './contexts/Context';
import DriverContext from './contexts/DriverContext';
import ScriptContext from './contexts/ScriptContext';
import ScaffoldContext from './contexts/ScaffoldContext';
import { Argv, Arguments, BeemoTool, Execution, BeemoPluginRegistry, BeemoConfig } from './types';

export default class Beemo {
  argv: Argv;

  moduleRoot: string = '';

  tool: BeemoTool;

  workspacePaths: string[] = [];

  constructor(argv: Argv, binName?: string, tool?: BeemoTool) {
    this.argv = argv;

    // eslint-disable-next-line global-require
    const { version } = require('../package.json');

    this.tool =
      tool ||
      new Tool<BeemoPluginRegistry, BeemoConfig>(
        {
          appName: 'beemo',
          appPath: path.join(__dirname, '..'),
          configBlueprint: this.getConfigBlueprint(),
          configName: binName,
          scoped: true,
        },
        argv,
      );

    this.tool.debug('Using beemo v%s', version);

    // Immediately load config and plugins
    this.tool.registerPlugin('driver', Driver).initialize();

    // Set footer after messages have been loaded
    this.tool.options.footer = `\nðŸ¤–  ${this.tool.msg('app:poweredBy', { version })}`;
  }

  /**
   * Register global options within our CLI application.
   */
  bootstrapCLI(app: YargsApp) {
    CLI.registerGlobalOptions(app, this.tool);
  }

  /**
   * If the configure module has an index export that is a function,
   * execute it with the current tool instance.
   */
  bootstrapConfigModule() {
    this.tool.debug('Bootstrapping configuration module');

    const moduleRoot = this.getConfigModuleRoot();
    const indexPath = path.join(moduleRoot, 'index.js');

    if (!fs.existsSync(indexPath)) {
      this.tool.debug('No index.js file detected, aborting bootstrap');

      return this;
    }

    // eslint-disable-next-line
    const bootstrap = require(indexPath);
    const isFunction = typeof bootstrap === 'function';

    this.tool.debug.invariant(isFunction, 'Executing bootstrap function', 'Found', 'Not found');

    if (isFunction) {
      bootstrap(this.tool);
    }

    return this;
  }

  /**
   * Create a configuration file for the specified driver names.
   */
  async createConfigFiles(
    args: Arguments,
    primaryDriver: string,
    additionalDrivers: string[] = [],
  ): Promise<any> {
    const { tool } = this;
    const driver = tool.getPlugin('driver', primaryDriver);
    const context = this.prepareContext(new DriverContext(args, driver));

    additionalDrivers.forEach(driverName => {
      context.addDriverDependency(tool.getPlugin('driver', driverName));
    });

    tool.debug('Running with %s driver(s)', [primaryDriver, ...additionalDrivers].join(', '));

    return this.startPipeline(context)
      .pipe(new ConfigureRoutine('config', tool.msg('app:configGenerate')))
      .run(primaryDriver);
  }

  /**
   * Define the blueprint for Beemo configuration.
   */
  getConfigBlueprint(): Blueprint {
    return {
      config: shape({
        cleanup: bool(false),
        parallel: bool(true),
      }),
      driver: shape({
        concurrency: number(),
        priority: bool(true),
      }),
      module: process.env.BEEMO_CONFIG_MODULE
        ? string(process.env.BEEMO_CONFIG_MODULE)
        : string().required(),
    };
  }

  /**
   * Validate the configuration module and return its absolute path.
   */
  getConfigModuleRoot(): string {
    if (this.moduleRoot) {
      return this.moduleRoot;
    }

    const { tool } = this;
    const { module } = tool.config;

    tool.debug('Locating configuration module root');

    if (!module) {
      throw new Error(tool.msg('errors:moduleConfigMissing'));
    }

    // Allow for local development
    if (module === '@local') {
      tool.debug('Using %s configuration module', chalk.yellow('@local'));

      this.moduleRoot = process.cwd();

      return this.moduleRoot;
    }

    // Reference a node module
    const rootPath = path.join(process.cwd(), 'node_modules', module);

    if (!fs.existsSync(rootPath)) {
      throw new Error(tool.msg('errors:moduleMissing', { module }));
    }

    tool.debug('Found configuration module root path: %s', chalk.cyan(rootPath));

    this.moduleRoot = rootPath;

    return rootPath;
  }

  /**
   * Return a list of absolute paths for Yarn or Lerna workspaces.
   */
  getWorkspacePaths(): string[] {
    if (this.workspacePaths.length > 0) {
      return this.workspacePaths;
    }

    const { workspaces } = this.tool.package;
    const root = this.tool.options.workspaceRoot || this.tool.options.root;
    const paths = [];

    if (workspaces) {
      if (Array.isArray(workspaces)) {
        paths.push(...workspaces);
      } else if (Array.isArray(workspaces.packages)) {
        paths.push(...workspaces.packages);
      }
    }

    const lernaPath = path.join(root, 'lerna.json');

    if (paths.length === 0 && fs.existsSync(lernaPath)) {
      const lerna = fs.readJsonSync(lernaPath);

      if (Array.isArray(lerna.packages)) {
        paths.push(...lerna.packages);
      }
    }

    this.workspacePaths = paths.map(workspace => path.join(root, workspace));

    return this.workspacePaths;
  }

  /**
   * Delete config files if a process fails.
   */
  handleCleanupOnFailure(code: number, context: DriverContext) {
    if (code === 0) {
      return;
    }

    // Must not be async!
    if (Array.isArray(context.configPaths)) {
      context.configPaths.forEach(config => {
        fs.removeSync(config.path);
      });
    }
  }

  /**
   * Execute all routines for the chosen driver.
   */
  async executeDriver(
    args: Arguments,
    driverName: string,
    parallelArgv: Argv[] = [],
  ): Promise<Execution[]> {
    const { tool } = this;
    const driver = tool.getPlugin('driver', driverName);
    const context = this.prepareContext(new DriverContext(args, driver, parallelArgv));

    // Delete config files on failure
    if (tool.config.config.cleanup) {
      tool.on(
        'exit',
        /* istanbul ignore next */ code => this.handleCleanupOnFailure(code, context),
      );
    }

    // Set workspace related properties
    context.workspaceRoot = tool.options.workspaceRoot || tool.options.root;
    context.workspaces = this.getWorkspacePaths();

    tool.emit(`${driverName}.init-driver`, [driver, context]);
    tool.debug('Running with %s driver', driverName);

    return this.startPipeline(context)
      .pipe(new ConfigureRoutine('config', tool.msg('app:configGenerate')))
      .pipe(new ExecuteDriverRoutine('driver', tool.msg('app:driverExecute')))
      .pipe(new CleanupRoutine('cleanup', tool.msg('app:cleanup')))
      .run(driverName);
  }

  /**
   * Run a script found within the configuration module.
   */
  async executeScript(args: Arguments, scriptName: string): Promise<Execution> {
    const { tool } = this;
    const context = this.prepareContext(new ScriptContext(args, scriptName));

    tool.emit(`${scriptName}.init-script`, [scriptName, context]);
    tool.debug('Running with %s script', scriptName);

    return this.startPipeline(context)
      .pipe(new ExecuteScriptRoutine('script', tool.msg('app:scriptExecute', { name: scriptName })))
      .run(scriptName);
  }

  /**
   * Prepare the context object by setting default values for specific properties.
   */
  prepareContext<T extends Context>(context: T): T {
    context.argv = this.argv;
    context.moduleRoot = this.getConfigModuleRoot();
    context.root = this.tool.options.root;

    return context;
  }

  /**
   * Run the scaffold process to generate templates.
   */
  async scaffold(args: Arguments, generator: string, action: string): Promise<any> {
    const { tool } = this;
    const context = this.prepareContext(new ScaffoldContext(args, generator, action));

    tool.emit(`${tool.options.appName}.scaffold`, [context, generator, action]);
    tool.debug('Running scaffold command');

    return this.startPipeline(context)
      .pipe(new ScaffoldRoutine('scaffold', tool.msg('app:scaffoldGenerate')))
      .run();
  }

  /**
   * Setup and start a fresh pipeline.
   */
  startPipeline<T extends Context>(context: T): Pipeline<T, BeemoTool> {
    // Make the tool available to all processes
    process.beemo = {
      context,
      tool: this.tool,
    };

    return new Pipeline(this.tool, context);
  }
}
