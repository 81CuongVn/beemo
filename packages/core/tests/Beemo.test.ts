import path from 'path';
import fs from 'fs-extra';
import optimal from 'optimal';
import Beemo from '../src/Beemo';
import Context from '../src/contexts/Context';
import DriverContext from '../src/contexts/DriverContext';
// @ts-ignore
import bootstrapIndex from '../../../tests';
import {
  getFixturePath,
  createDriverContext,
  createContext,
  createTestDriver,
  createTestTool,
} from '../../../tests/helpers';

jest.mock('fs-extra');

jest.mock(
  '@boost/core/lib/Pipeline',
  () =>
    function PipelineMock() {
      // @ts-ignore
      this.pipe = jest.fn(() => this);
      // @ts-ignore
      this.run = jest.fn(() => this);
    },
);

jest.mock('../../../tests', () => jest.fn());

const root = path.join(__dirname, '../../../tests');

describe('Beemo', () => {
  let beemo: Beemo;
  let onSpy: jest.Mock;
  const args = { _: [], $0: '' };

  beforeEach(() => {
    beemo = new Beemo(['foo', 'bar'], '', createTestTool());
    beemo.moduleRoot = root;
    beemo.tool.options.root = root;

    // Stop `exit` event from firing
    onSpy = jest.fn();
    beemo.tool.on = onSpy;

    (fs.existsSync as jest.Mock).mockReset();
  });

  it('sets argv', () => {
    expect(beemo.argv).toEqual(['foo', 'bar']);
  });

  describe('bootstrapConfigModule()', () => {
    beforeEach(() => {
      (bootstrapIndex as jest.Mock).mockReset();
      beemo.tool.config.module = '@local';
    });

    it('does nothing if no index file', () => {
      (fs.existsSync as jest.Mock).mockImplementation(() => false);

      beemo.bootstrapConfigModule();

      expect(bootstrapIndex).not.toHaveBeenCalled();
    });

    it('calls bootstrap with tool if index exists', () => {
      (fs.existsSync as jest.Mock).mockImplementation(() => true);

      beemo.bootstrapConfigModule();

      expect(bootstrapIndex).toHaveBeenCalledWith(beemo.tool);
    });
  });

  describe('createConfigFiles()', () => {
    beforeEach(() => {
      // @ts-ignore
      beemo.tool.getPlugin = (type, name) => createTestDriver(name);
    });

    it('triggers `init-driver` event with context', async () => {
      const spy = jest.spyOn(beemo.tool, 'emit');

      await beemo.createConfigFiles(args, 'foo');

      expect(spy).toHaveBeenCalledWith('foo.init-driver', [
        expect.objectContaining({ name: 'foo' }),
        expect.objectContaining({
          argv: ['foo', 'bar'],
          driverName: 'foo',
        }),
      ]);
    });

    it('passes context to pipeline', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');

      await beemo.createConfigFiles(args, 'foo');

      expect(spy).toHaveBeenCalledWith(
        expect.objectContaining({
          primaryDriver: expect.objectContaining({
            name: 'foo',
          }),
        }),
      );
    });

    it('passes multiple drivers', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');

      await beemo.createConfigFiles(args, 'foo', ['bar', 'baz']);

      expect(spy).toHaveBeenCalled();
      expect(spy.mock.calls[0][0].drivers.size).toBe(3);
    });
  });

  describe('getConfigBlueprint()', () => {
    it('errors if no module', () => {
      expect(() => {
        optimal({}, beemo.getConfigBlueprint());
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if module is defined with env var', () => {
      process.env.BEEMO_CONFIG_MODULE = 'test-boost';

      let opts: any = {};

      expect(() => {
        opts = optimal({}, beemo.getConfigBlueprint());
      }).not.toThrowError();

      expect(opts.module).toBe('test-boost');

      process.env.BEEMO_CONFIG_MODULE = '';
    });
  });

  describe('getConfigModuleRoot()', () => {
    beforeEach(() => {
      beemo.moduleRoot = '';
    });

    it('errors if no module name', () => {
      beemo.tool.config.module = '';

      expect(() => {
        beemo.getConfigModuleRoot();
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if a fake and or missing node module', () => {
      beemo.tool.config.module = 'beemo-this-should-never-exist';

      expect(() => {
        beemo.getConfigModuleRoot();
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if workspace path not found', () => {
      (fs.existsSync as jest.Mock).mockImplementation(() => false);

      beemo.workspacePaths = [path.join(process.cwd(), 'packages')];
      beemo.tool.config.module = '@local/foo';

      expect(() => {
        beemo.getConfigModuleRoot();
      }).toThrowErrorMatchingSnapshot();
    });

    it('returns cwd if using @local', () => {
      beemo.tool.config.module = '@local';

      expect(beemo.getConfigModuleRoot()).toBe(process.cwd());
      expect(beemo.moduleRoot).toBe(process.cwd());
      expect(beemo.getConfigModuleRoot()).toBe(beemo.moduleRoot);
    });

    it('returns node module path', () => {
      (fs.existsSync as jest.Mock).mockImplementation(() => true);

      beemo.tool.config.module = 'boost';

      const rootPath = path.join(process.cwd(), 'node_modules/boost');

      expect(beemo.getConfigModuleRoot()).toBe(rootPath);
      expect(beemo.moduleRoot).toBe(rootPath);
      expect(beemo.getConfigModuleRoot()).toBe(beemo.moduleRoot);
    });
  });

  describe('getWorkspacePaths()', () => {
    it('returns empty array for no workspaces', () => {
      beemo.tool.package = { name: '' };

      expect(beemo.getWorkspacePaths()).toEqual([]);
    });

    it('returns empty array for non-array workspaces', () => {
      beemo.tool.package = { name: '', workspaces: true };

      expect(beemo.getWorkspacePaths()).toEqual([]);
    });

    it('returns workspaces from package.json', () => {
      beemo.tool.package = { name: '', workspaces: ['packages/*'] };

      const paths = [path.join(root, 'packages/*')];

      expect(beemo.getWorkspacePaths()).toEqual(paths);
      expect(beemo.workspacePaths).toEqual(paths);
    });

    it('returns nohoist workspaces from package.json', () => {
      beemo.tool.package = {
        name: '',
        workspaces: {
          nohoist: [],
          packages: ['packages/*'],
        },
      };

      const paths = [path.join(root, 'packages/*')];

      expect(beemo.getWorkspacePaths()).toEqual(paths);
      expect(beemo.workspacePaths).toEqual(paths);
    });

    it('returns workspaces from lerna.json', () => {
      (fs.existsSync as jest.Mock).mockImplementation(() => true);
      (fs.readJsonSync as jest.Mock).mockImplementation(() => ({
        packages: ['packages/*'],
      }));

      beemo.tool.package = { name: '' };
      beemo.tool.options.workspaceRoot = getFixturePath('workspaces-lerna');

      const paths = [path.join(beemo.tool.options.workspaceRoot, 'packages/*')];

      expect(beemo.getWorkspacePaths()).toEqual(paths);
      expect(beemo.workspacePaths).toEqual(paths);
    });

    it('doesnt load lerna.json if workspaces are defined in package.json', () => {
      (fs.readJsonSync as jest.Mock).mockImplementation(() => ({
        packages: ['packages2/*'],
      }));

      beemo.tool.package = { name: '', workspaces: ['packages1/*'] };
      beemo.tool.options.workspaceRoot = getFixturePath('workspaces-lerna');

      expect(fs.existsSync as jest.Mock).not.toHaveBeenCalled();
      expect(beemo.getWorkspacePaths()).toEqual([
        path.join(beemo.tool.options.workspaceRoot, 'packages1/*'),
      ]);
    });
  });

  describe('handleCleanupOnFailure()', () => {
    let context: DriverContext;

    beforeEach(() => {
      context = createDriverContext();

      (fs.removeSync as jest.Mock).mockReset();
    });

    it('does nothing if exit code is 0', () => {
      beemo.handleCleanupOnFailure(0, context);

      expect(fs.removeSync).not.toHaveBeenCalled();
    });

    it('does nothing if no config paths', () => {
      beemo.handleCleanupOnFailure(1, context);

      expect(fs.removeSync).not.toHaveBeenCalled();
    });

    it('removes file for each config path', () => {
      context.configPaths = [{ driver: 'foo', path: 'foo' }, { driver: 'bar', path: 'bar' }];

      beemo.handleCleanupOnFailure(1, context);

      expect(fs.removeSync).toHaveBeenCalledWith('foo');
      expect(fs.removeSync).toHaveBeenCalledWith('bar');
    });
  });

  describe('executeDriver()', () => {
    beforeEach(() => {
      // @ts-ignore
      beemo.tool.getPlugin = () => ({ name: 'foo-bar' });
    });

    it('triggers `init-driver` event with context', async () => {
      const spy = jest.spyOn(beemo.tool, 'emit');

      await beemo.executeDriver(args, 'foo-bar');

      expect(spy).toHaveBeenCalledWith('foo-bar.init-driver', [
        expect.objectContaining({ name: 'foo-bar' }),
        expect.objectContaining({
          argv: ['foo', 'bar'],
          driverName: 'foo-bar',
        }),
      ]);
    });

    it('passes driver name and context to pipeline run', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');
      const pipeline = await beemo.executeDriver(args, 'foo-bar');

      expect(spy).toHaveBeenCalledWith(
        expect.objectContaining({
          argv: ['foo', 'bar'],
          driverName: 'foo-bar',
        }),
      );

      expect((pipeline as any).run).toHaveBeenCalledWith('foo-bar');
    });

    it('sets primary driver with context', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');

      await beemo.executeDriver(args, 'foo-bar');

      expect(spy).toHaveBeenCalledWith(
        expect.objectContaining({
          argv: ['foo', 'bar'],
          primaryDriver: expect.objectContaining({
            context: expect.objectContaining({
              driverName: 'foo-bar',
            }),
          }),
        }),
      );
    });

    it('passes parallelArgv to context', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');

      await beemo.executeDriver(args, 'foo', [['--foo'], ['bar']]);

      expect(spy).toHaveBeenCalledWith(
        expect.objectContaining({
          parallelArgv: [['--foo'], ['bar']],
        }),
      );
    });
  });

  describe('executeScript()', () => {
    it('triggers `init-script` event with context', async () => {
      const spy = jest.spyOn(beemo.tool, 'emit');

      await beemo.executeScript(args, 'foo-bar');

      expect(spy).toHaveBeenCalledWith('foo-bar.init-script', [
        'foo-bar',
        expect.objectContaining({
          argv: ['foo', 'bar'],
          scriptName: 'foo-bar',
        }),
      ]);
    });

    it('passes script name and context to pipeline run', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');
      const pipeline = await beemo.executeScript(args, 'foo-bar');

      expect(spy).toHaveBeenCalledWith(
        expect.objectContaining({
          argv: ['foo', 'bar'],
          scriptName: 'foo-bar',
        }),
      );

      expect((pipeline as any).run).toHaveBeenCalledWith('foo-bar');
    });
  });

  describe('prepareContext()', () => {
    it('sets extra props', () => {
      expect(beemo.prepareContext(new Context(args))).toEqual(
        expect.objectContaining({
          args,
          argv: ['foo', 'bar'],
          moduleRoot: root,
          root,
        }),
      );
    });
  });

  describe('startPipeline()', () => {
    beforeEach(() => {
      delete process.beemo;
    });

    it('sets beemo instance on process global', () => {
      expect(process.beemo).toBeUndefined();

      const context = createContext();
      context.args.foo = 123;
      context.args.bar = true;

      beemo.startPipeline(context);

      expect(process.beemo).toEqual({
        context,
        tool: beemo.tool,
      });
    });

    it('registers an exit listener if cleanup is true', async () => {
      beemo.tool.config.configure.cleanup = true;

      beemo.startPipeline(createContext());

      expect(onSpy).toHaveBeenCalledWith('exit', expect.any(Function));
    });

    it('doesnt register exit listener if cleanup is false', async () => {
      beemo.tool.config.configure.cleanup = false;

      beemo.startPipeline(createContext());

      expect(onSpy).not.toHaveBeenCalled();
    });
  });

  describe('scaffold()', () => {
    it('triggers `scaffold` event with context', async () => {
      const spy = jest.spyOn(beemo.tool, 'emit');

      await beemo.scaffold(args, 'gen', 'action');

      expect(spy).toHaveBeenCalledWith('beemo.scaffold', [
        expect.objectContaining({
          argv: ['foo', 'bar'],
        }),
        'gen',
        'action',
      ]);
    });

    it('passes context to pipeline', async () => {
      const spy = jest.spyOn(beemo, 'startPipeline');

      await beemo.scaffold(args, 'gen', 'action');

      expect(spy).toHaveBeenCalledWith(
        expect.objectContaining({
          generator: 'gen',
          action: 'action',
        }),
      );
    });
  });
});
