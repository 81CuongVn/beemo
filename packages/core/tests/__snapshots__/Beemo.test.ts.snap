// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Beemo getConfigBlueprint() errors if no module 1`] = `"Invalid field \\"module\\". Field is required and must be defined."`;

exports[`Beemo getConfigModuleRoot() errors if a fake and or missing node module 1`] = `"Module beemo-this-should-never-exist defined in \\"beemo.module\\" could not be found."`;

exports[`Beemo getConfigModuleRoot() errors if no module name 1`] = `"Beemo requires a \\"beemo.module\\" property within your package.json. This property is the name of a module that houses your configuration files."`;
